### ISSUE TITLE
Explicit Registration

### ISSUE THREAD

**Issue by bclothier** [WaynePhillipsEA/twinbasic#94]
**Is your feature request related to a problem? Please describe.**
When creating an ActiveX dll, tB will automatically register the classes. However it doesn't handle the cleanup, which mean we can end up with several registration debris throughout the development cycle and that is known to cause headaches.

**Describe the solution you'd like**
Instead of doing it magically, be more explicit. For example, expose the GUID in the source code, so that when a new class is created, it will be auto-generated but after that, the user has the option of keeping the GUID or manually editing it. With more transparency in how the classes/interfaces are decorated, it will help enormously with managing registration data, especially for the development where the API may be constantly changing. 

It would be also nice if tB is able to keep track of all GUIDs, detect whether a GUID is no longer and with user's consent remove the entries from the registry. 

**Describe alternatives you've considered**
A possible alternative is to always generate a reg script that will perform the unregistration of all the class for every build. That way, the developer can execute the reg script to effectively revert the stale registration data, even if the source code & GUIDs has since changed. It's then a simple matter of re-registering using the latest build.

**Comment by vbRichClient** [comment]
I agree... an "AutoRegister after Dll-compilation" switch (defaulting to "Off"?) -
in the Project-Settings-Page would make a lot of sense.
(especially when these nice regfree usable GetFactory-FlatAPI-Exports are linked into tB-Dlls by default :-) ... 
then the tB-IDE won't need any registry-entries anymore, to consume and use its own Dlls).

**Comment by Greedquest** [comment]
Do you need to register anything if you just want to import your DLL into VBA by adding a reference rather than using CreateObject? Will early bound code work without registration?

**Comment by wqweto** [comment]
> Will early bound code work without registration?

Registration has two parts -- interfaces registration point to a typelib (separate TLB or embedded in a DLL/EXE) and coclasses registration point to a DLL with DllGetClassObject export or EXE (no export here).

At runtime coclasses registration is used when instantiating classes with New (maps CLSID -> DLL) or CreateObject (maps ProgID -> DLL). At runtime interfaces registration is used only in multi-threaded scenarios so unless you go multi-threaded the typelib is not used.

So yes, compiled early bound code will very much work without typelib registration and regfree instantiation can be used without typelib registration too.

For interpreted early bound client code (VBA code that uses the component in the IDE) it's not possible to skip the typelib registration IMO.

**Comment by Kr00l** [comment]
IMO the guids should remain "managed" by tB. Otherwise it increases DLL hell and binary incompatibilities. One could take a DLL source code, change the interfaces and compile with same guids as the original one.

**Comment by bclothier** [comment]
Yet, we can still arrive to DLL hell via the automatic registration. It has happened before. 

With the explicit registration, the impetus is on the user to honor the COM conventions. The user has the freedom of retaining the same GUID but extending the interface... proven that it's compatible with the published version. If tB does it all behind the scene, then there's no way of knowing that a GUID is being preserved or being changed and that can spell bad news for distributing a new version of COM components. 

No, those who want to use COM must understand COM and adhere to the conventions. A compiler cannot cover all the edge cases, IMO.

**Comment by Kr00l** [comment]
Will there be a version compatibility in tB?
Because setting to "Binary Compatibility" it will automatically re-use the GUIDS compared to the select file and will guard by any breaks.
![image](https://user-images.githubusercontent.com/33756666/116406365-de83ac80-a830-11eb-87fd-b7e1ff5eb229.png)

**Comment by WaynePhillipsEA** [comment]
@Kr00l yes, this is planned, though I don't think we actually have a github issue for it yet.  

At the moment, GUIDs (coclass, IIDs, etc) are generated at compile time, if they are not explicitly set via attributes.  This is effectively the same as 'No compatibility'.  Project-persisted (cached) GUIDs will be supported soon, and then we can start refining to match VB's three settings.

**Comment by Kr00l** [comment]
@WaynePhillipsEA thanks. How does the GUID attribute works? This is already in place? And new feature, not existent in VB6?

Another question:
Is the GUID generation time-based or random?

**Comment by DaveInCaz** [comment]
Using attributes for GUIDs seems like a great substitute for VB6-style binary compatibility to me.  I've used similar features from C# and it seems to be extremely reliable and much easier than a cmp.dll to actually manage.  

I guess the one downside is that attributes and GUIDs can look "scary" to some programmers and that could be an impediment.

**Comment by DaveInCaz** [comment]
Related question - is COM going to be the mechanism for "linking" two DLLs that were both created using tB? If not than in many scenarios a COM interface wouldn't even be needed.

**Comment by WaynePhillipsEA** [comment]
Currently  COM is the only method of linking two DLLs, however standard DLL support is also planned (non-ActiveX), where functions can be exposed via standard DLL exports.  See #22

**Comment by DaveInCaz** [comment]
@WaynePhillipsEA  I'm just curious about this scenario... say its an EXE project which has 2 classes, `A` and `B`. You can obviously call `objA.Method1` from a method in `B` if you want. Is that using COM internally (I never thought so for VB6)?  Or is it more like a C++ DLL method which has mangled names that also include the class?

**Comment by WaynePhillipsEA** [comment]
No the call itself wouldn't use nor need any COM functionality.  The call would be made directly via the objects vtable, which the compiler knows the layout of through type-information that is built up during parsing/resolving (and which may or may not be in an actual exposed type library).

**Comment by DaveInCaz** [comment]
@WaynePhillipsEA in that case when you are building an EXE project which includes one or more DLLs, would the classes in the DLLs be "linked" in the same way?  The advantages there would be: 

- no registration concepts required whatsoever,
- very simple deployment is implicit (just copy everything to the same folder)
- possible performance boost (as compared to COM calls?)

One could argue that in such a case, why have DLLs at all? (Actually, "static linking" into a single EXE would be an interesting approach also) but there are many cases where the DLLs are meant for reuse.

Where COM registration is valuable, IMO, is not when you intend to consume the DLLs yourself, but more like where you are publishing an API (or controls) that you want your users to be able to consume by COM. But this is not always needed and the extra complications / overheads were (again IMO) quite painful in the VB6 days.
